src/hazmat/sieve.rs

```diff
@@ -11,21 +11,46 @@ use rand_core::CryptoRngCore;
 use crate::hazmat::precomputed::{SmallPrime, LAST_SMALL_PRIME, RECIPROCALS, SMALL_PRIMES};
 use crate::traits::SieveFactory;
 
-/// Returns a random odd integer with given bit length
+/// Set bits in the returned number.
+#[derive(Debug, Clone, Copy)]
+pub enum SetBits {
+    /// No additional bits set.
+    None,
+    /// Set the most significant bit.
+    Msb,
+    /// Set two most significant bits.
+    ///
+    /// This is useful in the RSA case because a product of two such numbers will have a guaranteed bit size.
+    TwoMsb,
+}
+
+/// Returns a random odd integer up to the given bit length
 /// (that is, with both `0` and `bit_length-1` bits set).
 ///
+/// Additionally, some bits can be set depending on the `set_bits` value.
+///
 /// *Panics*: if the `bit_length` is bigger than the bits available in the `Integer`, e.g. 37 for a
 /// `U32`.
-pub fn random_odd_integer<T: Integer + RandomBits>(rng: &mut impl CryptoRngCore, bit_length: NonZeroU32) -> Odd<T> {
+pub fn random_odd_integer<T: Integer + RandomBits>(
+    rng: &mut impl CryptoRngCore,
+    bit_length: NonZeroU32,
+    set_bits: SetBits,
+) -> Odd<T> {
     let bit_length = bit_length.get();
 
     let mut random = T::random_bits(rng, bit_length);
     // Make it odd
     random.set_bit_vartime(0, true);
 
-    // Make sure it's the correct bit size
     // Will not overflow since `bit_length` is ensured to be within the size of the integer.
```

Comment:
This comment is not correct: we're not actually checking this.

Reply:
`T::random_bits()` does. I'll add it to the comment.